File: 0144 Binary Tree Preorder Traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public List<Integer> preorderTraversal(TreeNode root) {
        List<Integer> res = new ArrayList<>();
        preorder(root, res);
        return res;
    }
    public void preorder(TreeNode root, List<Integer> res) {
        if (root == null) return;
        res.add(root.val);
        preorder(root.left, res);
        preorder(root.right, res);
    }
}

/* Iterative solution using Stack
// root -> left -> right => In stack: push right->left->root (LIFO)
class Solution {
    public List<Integer> preorderTraversal(TreeNode root) {
        List<Integer> res = new ArrayList<>();
        Stack<TreeNode> st = new Stack<TreeNode>();
        if (root != null) {
            st.push(root);
            while (!st.isEmpty()) {
                TreeNode n = st.pop();
                if (n.right != null) st.push(n.right);
                if (n.left != null) st.push(n.left);
                res.add(n.val);
            }
        }
        return res;
    }
}
*/

File: 0144 Binary Tree Preorder Traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def preorderTraversal(self, root: TreeNode) -> List[int]:
        res = []
        self.preorder(root, res)
        return res
        
    def preorder(self, root: TreeNode, res: List[int]):
        if not root: return
        res.append(root.val)
        self.preorder(root.left, res)
        self.preorder(root.right, res)

'''
# root -> left -> right => In stack: push right->left->root (LIFO)

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def preorderTraversal(self, root: TreeNode) -> List[int]:
        res = []
        stack = []
        if root:
            stack.append(root)
            while stack:
                n = stack.pop()
                if n.right: stack.append(n.right)
                if n.left: stack.append(n.left)
                res.append(n.val)
        return res

'''